crates/shared/src/price_estimation/sanitized.rs

```diff
@@ -0,0 +1,331 @@
+use super::{Estimate, PriceEstimating, PriceEstimationError, Query};
+use crate::bad_token::{BadTokenDetecting, TokenQuality};
+use crate::price_estimation::gas::GAS_PER_WETH_UNWRAP;
+use anyhow::Result;
+use model::order::BUY_ETH_ADDRESS;
+use primitive_types::H160;
+use std::collections::{HashMap, HashSet};
+use std::sync::Arc;
+
+/// Verifies that buy and sell tokens are supported and handles
+/// ETH as buy token appropriately.
+pub struct SanitizedPriceEstimator<T: PriceEstimating> {
+    inner: T,
+    bad_token_detector: Arc<dyn BadTokenDetecting>,
+    native_token: H160,
+}
+
+type EstimationResult = Result<Estimate, PriceEstimationError>;
+
+enum EstimationProgress<'a> {
+    TrivialSolution(EstimationResult),
+    AwaitingEthEstimation(&'a Query),
+    AwaitingErc20Estimation,
+}
+
+impl<T: PriceEstimating> SanitizedPriceEstimator<T> {
+    pub fn new(
+        inner: T,
+        native_token: H160,
+        bad_token_detector: Arc<dyn BadTokenDetecting>,
+    ) -> Self {
+        Self {
+            inner,
+            native_token,
+            bad_token_detector,
+        }
+    }
+
+    async fn get_token_quality_errors(
+        &self,
+        queries: &[Query],
+    ) -> HashMap<H160, PriceEstimationError> {
+        let mut token_quality_errors: HashMap<H160, PriceEstimationError> = Default::default();
+        let mut checked_tokens = HashSet::<H160>::default();
+
+        for token in queries
+            .iter()
+            .copied()
+            .flat_map(|query| [query.buy_token, query.sell_token])
+        {
+            if checked_tokens.contains(&token) {
+                continue;
+            }
+
+            match self.bad_token_detector.detect(token).await {
```

Comment:
nanonit: in theory these could be done in parallel (likely premature optimisation but maybe something to comment as a note?)

Reply:
Done.